src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -100,6 +104,95 @@ def pileupSizeTask(self):
             self.logger.exception(msg)
         return cmsDict, cusDict
 
+    def partialPileupTask(self, rseExpression):
+        """
+        Execute partial pileup placement according to the following logic:
+
+        - get rucio DIDs (datasets, in CMS name blocks) for our pileup document
+        - get portion of DIDs based on ceil(containerFraction * num_rucio_datasets)
+        - create new container DID in Rucio via Rucio wrapper createContainer API
+          which calls rucio CLI add_container
+        - update MSPileup JSON with
+          - create new custom Name as pileup+extention
+          - add new transition record and update MSPileup document
+        - we call attachDIDs Rucio wrapper API with our set of DIDs and rses from pileup document
+        - create new rules for custom DID via provided rseExpression
+        - set expiration date (to be 24h) for already existing ruleIds from pileup document
+        - add new ruleIds to pileup document
+
+        :param rseExpression: string, Rucio rse expression, see
+        https://rucio.cern.ch/documentation/started/concepts/rse_expressions/
+        :return: None
+        """
+        spec = {'containerFraction': {"$gt": 0}}
+        docs = self.mgr.getPileup(spec)
+        for doc in docs:
+            self.logger.info("partialPileupTask process %s", doc)
+            fraction = doc['containerFraction']
+
+            # check previous fraction value and proceed only if has changed
+            lastFraction = 0
+            transition = doc.get('transition', [])
+            if transition:
+                lastFraction = transition[-1]['containerFraction']
+            if lastFraction == fraction:
+                # do nothing since there is no transition, i.e. skip this record
+                continue
+
+            # get rucio DIDs (datasets, in CMS name blocks) for our pileup document as following:
+            # - when container fraction is decreasing: new container is a subset of customName
+            # - when container fraction is increasing: new container is a superset of customName + subset of pileupName
+
+            # by default we assume it is decreasing
+            pname = doc['customName']
+            blockNames = self.rucioClient.getBlocksInContainer(pname)
+
+            # but if it increasing
+            if lastFraction < fraction:
+                pname = doc['pileupName']
+                blockNames += self.rucioClient.getBlocksInContainer(pname)
+
+            # get portion of DIDs based on ceil(containerFraction * num_rucio_datasets)
+            portion = blockNames[:math.ceil(fraction * len(blockNames))]
```

Comment:
I made a change to the code as if/else block make it clear, and added appropriate comment

Reply:
Please see my comment made against L155 (above). We should not blindly pick block names, otherwise we risk having to stage data while other data is already on Disk.

Reply:
My gist has been updated:
https://gist.github.com/amaltaro/b4f9bafc0b58c10092a0735c635538b5#logic-for-increasingdecreasing-container-fraction
in case you need to apply corrections here.